tests/5.0/allocate/test_allocate.c

```diff
@@ -0,0 +1,47 @@
+//===----------------- test_allocate.c ------------------------------------===//
+//
+// OpenMP API Version 5.0 Nov 2018
+//
+// Tests the allocate directive. A variable 'x' is allocated using 
+// the omp_alloc call. If no clause is specified then the memory allocator s
+// pecified by the def-allocator-var ICV will be used. The tests 
+// checks that the values were written correctly, and then frees the memory.
+//
+//===----------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1024
+
+int test_allocate() {
+  int errors = 0;
+  int* x;
+
+#pragma omp allocate(x) 
+  x = (int *) omp_alloc(N*sizeof(int), omp_null_allocator);
```

Comment:
I think I don't understand this example. I do understand:
* `int *x; x = (int*) omp_alloc( .... );` — This allocates explicitly with a routine call, which has to be freed by the user (→ omp_free).
* `int A[10], *y; #pragma omp allocate(A, y)` – This allocates the memory - and it is automatically freed at the end of the scope of the variable [directive and variable are in the same scope]. (Though, I am not 100% sure about the 'y' semantic; I think it allocates a scalar pointer - but I might be wrong. All examples I saw for the 'allocate' directive use constant arrays [but ↓].)

In any case, I fail to understand why there is the `allocate` directive _and_ an `omp_alloc` for the same variable.

I think the only exception where the `omp allocate` does not stand on its own is for Fortran with allocatables/pointers, for which the directive line is followed by a Fortran ALLOCATE statement. But to avoid semantic overloading – and also to avoid issues in the compiler about whether the directive belongs to the declaration or execution part, the later was changed by introducing the new `allocators` construct in OpenMP 5.2 (and deprecating the `allocate` directive for this purpose, only).

_Disclaimer: I might have misunderstood the OpenMP specification._

BTW: GCC: _'omp allocate' support exists as patch but it is not yet in GCC 12.  However, calling `omp_alloc` in GCC in target regions is fine. (I think since GCC 11, but I have not checked.)_

Reply:
Side note part two: GCC – additional minor comments regarding GCC's implementation status:
* `uses_allocators` on `target` is not supported, but `requires dynamic_allocators` is fulfilled (also in GCC 11; but since GCC 12 it is accepted by turning into a no op). – Thus, the omp_alloc etc. should work since GCC 11 (also in `target`). But GCC will choke on the `uses_allocators`
* GCC _does_ support the `allocate` clause as in `omp task private(a) allocate(a)` for C/C++ since GCC 11. And since GCC 12, it should support it also for Fortran and also OpenMP 5.1's `allocate (allocator (h), align (8): n2)` in C/C++.
  _I don't know whether you already have a testcase for the `allocate` clause or not._

Reply:
@tob2  This is a 5.0 test. Allocate directive just declares the access traits required, while omp_alloc is where the allocation happens. I agree, allocate followed by omp_alloc for the same mem space might be an overkill (though completely legal). For the on device test, I am going to do the variation with the stack allocate (and that passes with LLVM, not checked with GCC yet). 
For my information, in the GCC implementation, if a regular malloc is used after the allocate directive, does it get intercepted to by omp runtime to verify the access traits of the memory and allocate memory accordingly?

Reply:
@spophale In my understanding, Allocate directive in C/C++ is a stand-alone directive, not associated with a statement. Therefore, actual memory allocation happens in the Allocate directive, which means that this example has duplicated allocations for the same variable, which is an overkill but still legal.

Reply:
Aah, I see. From the language- "the storage for each list item that appears in the directive will be provided by an allocation through a memory allocator" it was not very clear to me that the allocation happens via the allocate directive or we need to use an explicit **memory allocator**. It makes sense that allocate provides the memory.

Reply:
Thanks @tob2 and @seyonglee !

Reply:
> Allocate directive just declares the access traits required, while ... is where the allocation happens.

That's actually the case — but only for ALLOCATABLE/POINTER variables in Fortran:
```F90
!$omp allocate allocator(my_alloc)  ! this usage is deprecated - and replaced by the 'allocators' construct in OpenMP 5.1
allocate (x(N))   ! uses the Fortran allocate statement.
```